File: analysis/import/SCMLogParser/src/main/java/de/maibornwolff/codecharta/importer/scmlogparser/input/metrics/HighlyCoupledFiles.java
package de.maibornwolff.codecharta.importer.scmlogparser.input.metrics;

import de.maibornwolff.codecharta.importer.scmlogparser.input.Commit;
import de.maibornwolff.codecharta.importer.scmlogparser.input.Modification;

import java.util.HashMap;
import java.util.Map;

public class HighlyCoupledFiles implements Metric {
    private static final double HIGH_COUPLING_VALUE = .35;
    private static final long MIN_NO_COMMITS_FOR_HIGH_COUPLING = 5L;

    private final Map<String, Long> simultaneouslyCommitedFiles = new HashMap<>();
    private long numberOfCommits = 0;

    @Override
    public String description() {
        return "Highly Coupled Files: Number of highly coupled files (35% times modified the same time) with this file.";
    }

    @Override
    public String metricName() {
        return "highly_coupled_files";
    }

    @Override
    public Number value() {
        return simultaneouslyCommitedFiles.values().stream()
                .filter(this::isHighlyCoupled)
                .count();
    }

    private boolean isHighlyCoupled(long val) {
        if (numberOfCommits >= MIN_NO_COMMITS_FOR_HIGH_COUPLING) {
            return (double) val / (double) numberOfCommits > HIGH_COUPLING_VALUE;
        }

        return false;
    }

    @Override
    public void registerCommit(Commit commit) {
        numberOfCommits++;
        commit.getModifications()
                .forEach(
                        mod -> simultaneouslyCommitedFiles.merge(mod.getFilename(), 1L, (x, y) -> x + y)
                );
    }

    @Override
    public void registerModification(Modification modification) {
        // delete this file
        simultaneouslyCommitedFiles.remove(modification.getFilename());
    }
}

File: analysis/import/SCMLogParser/src/main/kotlin/de/maibornwolff/codecharta/importer/scmlogparser/input/metrics/HighlyCoupledFiles.kt
package de.maibornwolff.codecharta.importer.scmlogparser.input.metrics

import de.maibornwolff.codecharta.importer.scmlogparser.input.Commit
import de.maibornwolff.codecharta.importer.scmlogparser.input.Modification
import de.maibornwolff.codecharta.model.AttributeType
import de.maibornwolff.codecharta.model.Edge

// TODO track all renames to preserve coupling, preprocessing to prevent exponential growth
// file named A -> renamed to B, created named file A (bug)
class HighlyCoupledFiles : Metric {

    private var fileName: String = ""
    private var numberOfCommits: Long = 0
    private var commits = mutableListOf<Commit>()
    private val simultaneouslyCommittedFiles = mutableMapOf<String, Int>()

    override fun description(): String {
        return "Highly Coupled Files: Number of highly coupled files (35% times modified the same time) with this file."
    }

    override fun metricName(): String {
        return "highly_coupled_files"
    }

    override fun edgeMetricName(): String {
        return "temporal_coupling"
    }

    override fun value(): Number {
        evaluateIfNecessary()

        return simultaneouslyCommittedFiles.values
            .filter { isHighlyCoupled(it) }
            .count()
            .toLong()
    }

    override fun getEdges(): List<Edge> {
        evaluateIfNecessary()

        return simultaneouslyCommittedFiles
            .filter { isHighlyCoupled(it.value) }
            .map { (coupledFile, value) ->
                Edge(fileName, coupledFile, mapOf(edgeMetricName() to value.toDouble() / numberOfCommits.toDouble()))
            }
    }

    private fun evaluateIfNecessary() {
        if (simultaneouslyCommittedFiles.isNotEmpty()) return

        commits.forEach { commit ->
            commit.modifications
                    .filter { it.currentFilename != fileName }
                    .forEach { simultaneouslyCommittedFiles.merge(it.currentFilename, 1) { x, y -> x + y } }
        }
    }

    private fun isHighlyCoupled(value: Int): Boolean {
        return if (value >= MIN_NO_COMMITS_FOR_HIGH_COUPLING) {
            value.toDouble() / numberOfCommits.toDouble() > HIGH_COUPLING_VALUE
        } else false
    }

    override fun registerCommit(commit: Commit) {
        numberOfCommits++
        commits.add(commit)
    }

    override fun registerModification(modification: Modification) {
        fileName = modification.currentFilename
    }

    override fun edgeAttributeType(): AttributeType? {
        return AttributeType.absolute
    }

    companion object {
        private const val HIGH_COUPLING_VALUE = .35
        private const val MIN_NO_COMMITS_FOR_HIGH_COUPLING = 5L
    }
}
